lib/api/model/submessage.dart

```diff
@@ -0,0 +1,397 @@
+  @JsonKey(includeToJson: true)
+  PollEventSubmessageType get type => PollEventSubmessageType.newOption;
+
+  final String option;
+  /// The index of last [option] added by the sender.
+  @JsonKey(name: 'idx')
+  final int latestOptionIndex;
+
+  PollNewOptionEventSubmessage({required this.option, required this.latestOptionIndex});
+
+  @override
+  factory PollNewOptionEventSubmessage.fromJson(Map<String, Object?> json) =>
+    _$PollNewOptionEventSubmessageFromJson(json);
+
+  @override
+  Map<String, Object?> toJson() => _$PollNewOptionEventSubmessageToJson(this);
+}
+
+/// A poll event when the question has been edited.
+@JsonSerializable(fieldRename: FieldRename.snake)
+class PollQuestionEventSubmessage extends PollEventSubmessage {
+  @override
+  @JsonKey(includeToJson: true)
+  PollEventSubmessageType get type => PollEventSubmessageType.question;
+
+  final String question;
+
+  PollQuestionEventSubmessage({required this.question});
+
+  @override
+  factory PollQuestionEventSubmessage.fromJson(Map<String, Object?> json) =>
+    _$PollQuestionEventSubmessageFromJson(json);
+
+  @override
+  Map<String, Object?> toJson() => _$PollQuestionEventSubmessageToJson(this);
+}
+
+/// A poll event when a vote has been cast or removed.
+@JsonSerializable(fieldRename: FieldRename.snake)
+class PollVoteEventSubmessage extends PollEventSubmessage {
+  @override
+  @JsonKey(includeToJson: true)
+  PollEventSubmessageType get type => PollEventSubmessageType.vote;
+
+  /// The key of the affected option.
+  ///
+  /// See [PollEventSubmessage.optionKey].
+  final String key;
+  @JsonKey(name: 'vote', unknownEnumValue: PollVoteOp.unknown)
+  final PollVoteOp op;
+
+  PollVoteEventSubmessage({required this.key, required this.op});
+
+  @override
+  factory PollVoteEventSubmessage.fromJson(Map<String, Object?> json) {
+    final result = _$PollVoteEventSubmessageFromJson(json);
+    // Crunchy-shell validation
+    final segments = result.key.split(',');
+    final [senderId, optionIndex] = segments;
+    if (senderId != 'canned') {
+      int.parse(senderId, radix: 10);
+    }
+    int.parse(optionIndex, radix: 10);
+    return result;
+  }
+
+  @override
+  Map<String, Object?> toJson() => _$PollVoteEventSubmessageToJson(this);
+}
+
+/// As in [PollVoteEventSubmessage.op].
+@JsonEnum(valueField: 'apiValue')
+enum PollVoteOp {
+  add(apiValue: 1),
+  remove(apiValue: -1),
+  unknown(apiValue: null);
+
+  const PollVoteOp({required this.apiValue});
+
+  final int? apiValue;
+
+  int? toJson() => apiValue;
+}
+
+class UnknownPollEventSubmessage extends PollEventSubmessage {
+  @override
+  @JsonKey(includeToJson: true)
+  PollEventSubmessageType get type => PollEventSubmessageType.unknown;
+
+  final Map<String, Object?> json;
+
+  UnknownPollEventSubmessage({required this.json});
+
+  @override
+  Map<String, Object?> toJson() => json;
+}
+
+/// As in [PollEventSubmessage.type].
+@JsonEnum(fieldRename: FieldRename.snake)
+enum PollEventSubmessageType {
```

Comment:
nit: let's put this right after PollEventSubmessage — it's closely tied to that base class (for example the `fromJson` there has a switch that's on exactly these values), and it provides the outline of what the subclasses will be.